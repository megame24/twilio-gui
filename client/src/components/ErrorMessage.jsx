import React from 'react';
import { PropTypes } from 'prop-types';
import { Message } from 'semantic-ui-react';

/**
 * ErrorMessage component
 */
export class ErrorMessage extends React.Component {
  _isMounted = false;

  state = { visible: false }

  /**
   * @returns {undefined}
   */
  componentDidMount() {
    this._isMounted = true;
  }

  /**
   * @returns {undefined}
   */
  componentWillUnmount() {
    this._isMounted = false;
  }

  /**
   * @returns {undefined}
   */
  handleDismiss = () => {
    this.setState({ visible: false });
  }

  /**
   * @returns {undefined}
   */
  dismissOnTimeout = () => {
    setTimeout(() => {
      if (this._isMounted) {
        this.setState({ visible: false });
      }
    }, 10000);
  }

  /**
   * component will receive props
   * @param {Object} nextProps
   * @returns {undefined}
   */
  componentWillReceiveProps(nextProps) {
    const { errors } = this.props;
    if (nextProps.errors.message && errors.time !== nextProps.errors.time) {
      this.setState({ visible: true });
      this.dismissOnTimeout();
    }
  }

  /**
   * @returns {undefined}
   */
  render() {
    const { errors } = this.props;
    const { visible } = this.state;
    return (
      visible && (
        <Message
          negative
          onDismiss={this.handleDismiss}
          header={errors.message}
        />
      )
    );
  }
}

ErrorMessage.propTypes = {
  errors: PropTypes.shape({
    message: PropTypes.string,
    time: PropTypes.instanceOf(Date)
  }),
};

ErrorMessage.defaultProps = {
  errors: {},
};

export default ErrorMessage;
