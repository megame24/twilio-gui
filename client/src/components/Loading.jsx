import React from 'react';
import { PropTypes } from 'prop-types';
import { Dimmer, Loader } from 'semantic-ui-react';

const Loading = ({ isLoading }) => (
  <Dimmer
    page
    active={isLoading}
  >
    <Loader size="medium" />
  </Dimmer>
);

Loading.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};

export default Loading;
