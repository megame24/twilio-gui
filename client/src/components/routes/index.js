import GuestRoute from './GuestRoute';
import UserRoute from './UserRoute';

export default {
  GuestRoute,
  UserRoute,
};
