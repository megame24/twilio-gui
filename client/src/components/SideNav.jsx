import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import contactActions from '../actions/contactActions';
import authActions from '../actions/authActions';
import socket from '../repositories/Socket';
import ErrorMessage from './ErrorMessage';

/**
 * SideNav component
 */
export class SideNav extends React.Component {
  /**
   * component did mount
   * @returns {undefined}
   */
  componentDidMount() {
    const { getContacts, contacts } = this.props;
    if (Object.keys(contacts).length < 1) {
      getContacts();
    }
  }

  /**
   * component will receive props
   * @param {Object} nextProps
   * @returns {undefined}
   */
  componentWillReceiveProps(nextProps) {
    const { getContacts, token } = this.props;
    if (nextProps.token !== token) {
      getContacts();
    }
  }

  /**
   * logout
   * @returns {undefined}
   */
  logout = () => {
    const { logout } = this.props;
    socket.disconnect();
    logout();
  }

  /**
   * @return {Function} JSX function
   */
  render() {
    const { contacts, errors } = this.props;
    errors.time = new Date();
    return (
      <div id="side-nav">
        <div className="errors">
          <ErrorMessage
            errors={errors}
          />
        </div>
        <Link to="/" className="new-message">
          <i className="ion-plus-circled" />
          New Message
        </Link>
        <a to="#" onClick={this.logout} className="logout">Logout</a>
        <div className="contact-list">
          <h2>Contacts</h2>
          <ul>
            {
              Object.values(contacts).map(contact => (
                <li key={contact.id}>
                  <Link to={`/contacts/${contact.id}`} className="contact">
                    <i className="ion-android-contact" />
                    <span>
                      {contact.name && <p className="name">{contact.name}</p>}
                      <p>{contact.phoneNumber}</p>
                    </span>
                    {
                      contact.notifications
                      && (
                        <span className="notifn">
                          <i className="ion-android-mail" />
                        </span>
                      )
                    }
                  </Link>
                </li>
              ))
            }
          </ul>
        </div>
      </div>
    );
  }
}

SideNav.propTypes = {
  logout: PropTypes.func.isRequired,
  token: PropTypes.string,
  getContacts: PropTypes.func.isRequired,
  contacts: PropTypes.shape({
    phoneNumber: PropTypes.string
  }),
  errors: PropTypes.shape({
    message: PropTypes.string
  }),
};

SideNav.defaultProps = {
  token: '',
  contacts: {},
  errors: {},
};

export const mapStateToProps = ({
  contacts: { contacts, errors },
  auth: { token }
}) => ({
  contacts,
  token,
  errors,
});

export default connect(mapStateToProps, {
  getContacts: contactActions.getContacts,
  logout: authActions.logout
})(SideNav);
