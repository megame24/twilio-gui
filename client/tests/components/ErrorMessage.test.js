import React from 'react';
import { shallow } from "enzyme";
import { ErrorMessage } from '../../src/components/ErrorMessage';

describe('Testing ErrorMessage component', () => {
  describe('Testing component methods', () => {
    const props = {
      errors: {
        message: 'error!!',
        time: new Date()
      }
    }
    const ErrorMessageComp = () => shallow(<ErrorMessage { ...props } />);
    it('Should set _isMounted to true when the component is mounted', () => {
      const ErrorMessageCompInstance = ErrorMessageComp().instance();
      expect(ErrorMessageCompInstance._isMounted).toBeTruthy();
    });
    it('Should set _isMounted to false when the component is unmounted', () => {
      jest.spyOn(ErrorMessage.prototype, 'componentWillUnmount');
      ErrorMessageComp().unmount();
      expect(ErrorMessage.prototype.componentWillUnmount.mock.calls.length).toEqual(1);
      expect(ErrorMessage.prototype._isMounted).toBeFalsy();
    });
    it('Should set visible to false when handleDismiss is called', async () => {
      const ErrorMessageCompInstance = ErrorMessageComp().instance();
      ErrorMessageCompInstance.state.visible = true;
      ErrorMessageCompInstance.handleDismiss();
      expect(ErrorMessageCompInstance.state.visible).toBeFalsy();
    });
    it('Should set visible to false when dismissOnTimeout is called and _isMounted is true', async () => {
      const ErrorMessageCompInstance = ErrorMessageComp().instance();
      ErrorMessageCompInstance.state.visible = true;
      await ErrorMessageCompInstance.dismissOnTimeout();
      expect(ErrorMessageCompInstance.state.visible).toBeFalsy();
    });
    it('Should do nothing when dismissOnTimeout is called and _isMounted is false', async () => {
      const ErrorMessageCompInstance = ErrorMessageComp().instance();
      ErrorMessageCompInstance.state.visible = true;
      ErrorMessageCompInstance._isMounted = false;
      await ErrorMessageCompInstance.dismissOnTimeout();
      expect(ErrorMessageCompInstance.state.visible).toBeTruthy();
    });
    it('Should set visible to true when componentWillReceiveProps is called', () => {
      const ErrorMessageCompt = ErrorMessageComp();
      const dismissOnTimeout = jest.spyOn(ErrorMessageCompt.instance(), 'dismissOnTimeout');
      dismissOnTimeout.mockReset();
      const newErrors = {
        message: 'new errors',
        time: new Date
      };
      ErrorMessageCompt.setProps({ errors: newErrors });
      expect(dismissOnTimeout.mock.calls.length).toEqual(1);
      expect(ErrorMessageCompt.instance().state.visible).toBeTruthy();
    });
    it('Should render as expected when visible is true', () => {
      const tree = ErrorMessageComp();
      tree.setState({ visible: true });
      expect(tree).toMatchSnapshot();
    });
    it('Should render as expected when visible is false', () => {
      const tree = ErrorMessageComp();
      tree.setState({ visible: false });
      expect(tree).toMatchSnapshot();
    });
  });
});
