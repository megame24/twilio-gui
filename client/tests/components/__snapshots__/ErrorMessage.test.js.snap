// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ErrorMessage component Testing component methods Should render as expected when visible is false 1`] = `""`;

exports[`Testing ErrorMessage component Testing component methods Should render as expected when visible is true 1`] = `
<Message
  header="error!!"
  negative={true}
  onDismiss={[Function]}
/>
`;
